### PR TITLE
Add css load checker to prevent duplicate loading of style sheet

### PR DESCRIPTION
**Pull request by bv-jason-dippel**
This adds the cssLoadChecker module, which is a means of preventing the duplicate loading of requested style sheets.

Research was done into the possibility of using the SAS client in bv-ui-core, but this would not work with the existing use case for this module as explained [here](https://github.com/bazaarvoice/bv-ui-core/issues/52#issuecomment-193389675).

### CODE REVIEW

File: lib/cssLoadChecker/index.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ * Function to check whether the className corresponds to
+ * a loaded style sheet
+ *
+ * @param {string} className - The class name used to check if the css file has loaded yet
+ *  (does not include the '.' out front, just the name)
+ * @return {boolean} True or false as to whether the style sheet has been loaded.
+ */
+var _checkForCss = function (className) {
+  var testDiv = _loadTestDiv(className);
+  var computedStyles = global.getComputedStyle(testDiv);
```
Comment (reason-bv) [inline lib/cssLoadChecker/index.js]: Not supported in IE8, so I think some kind of check is needed here.

File: lib/cssLoadChecker/index.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ */
+var _loadTestDiv = function (className) {
+  var testDiv;
+
+  testDiv = global.document.createElement('div');
+  testDiv.style.height = '0px';
+  testDiv.style.width = '0px';
+  testDiv.style.border = '0px';
+  testDiv.className = className;
+  global.document.body.appendChild(testDiv);
```
Comment (reason-bv) [inline lib/cssLoadChecker/index.js]: There is a `body` module in `bv-ui-core`. Better to use that.

**Comment by reason-bv** [comment]
So it seems to me that rather than doing what you are doing here, you'd be better off extending the loader to keep a record of what it loads, and pass in a flag to force reload, and just make sure it never tries to load stylesheets twice by default.

If you do have other use cases for checking computed styles to see what they are, then salvaging the core of this may still be helpful for other uses.

**Comment by sirugh** [comment]
We should at least keep the checker module separate from the loader. Requesting and parsing CSS are two separate concerns and should be captured here in this PR.

**Comment by tnunamak** [comment]
> So it seems to me that rather than doing what you are doing here, you'd be better off extending the loader to keep a record of what it loads, and pass in a flag to force reload, and just make sure it never tries to load stylesheets twice by default.

The PRS use case driving this involves the PRS waiting for a stylesheet that Firebird is responsible for loading. Unless multiple instances of the loader share state about what is loaded, we'll still need a way to recognize when CSS that we did not load is available.

**Comment by sirugh** [comment]
@tnunamak I do see what @reason-bv was saying though. There is already functionality to issue requests for resources, so part of this PR is unnecessary.

**Comment by tnunamak** [comment]
Lack of due diligence on my part, you're right.

**Comment by reason-bv** [comment]
So in theory we're all using a shared namespace, so there could be a trackedloader module that writes to that namespace, or the loader could take a configuration that tells it to write to a shared place in that namespace. We should be all about sharing state between apps when it helps us.

**Comment by bv-jason-dippel** [comment]
So in response to @reason-bv's comments, I've added some stuff to the loader module that keeps track of what URLs we've loaded files from and will only attempt to reload them if the options.forceLoad flag is passed.

I do believe we should keep the cssLoadChecker module in addition to this for the specific use case that brought these changes around. The PRS use case driving this does not have access to the URL firebird uses to load the CSS, so we must check whether the CSS has loaded yet or not by looking at the computed styles.

File: lib/cssLoadChecker/index.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    if (displayVal === 'none') {
+      result = true;
+    }
+
+    _cleanupTestDiv(testDiv);
+    return result;
+  }
+
+}
```
Comment (reason-bv) [inline lib/cssLoadChecker/index.js]: Semicolon.

File: lib/loader/index.js [hunk 4]
```diff
@@ -14,9 +14,13 @@
  */
 
 var global = require('../global');
+var namespacer = require('../namespacer');
 var doc = global.document;
 
 var DEFAULT_TIMEOUT = 10000;
+var SHARED_NAMESPACE_NAME = 'BV_LOADER_TRACKING';
```
Comment (reason-bv) [inline lib/loader/index.js]: I was thinking you'd take an optional namespace as a parameter / setting / something, so that only apps sharing a namespace share this.

For CSS it doesn't matter, but for JS we have the interesting situation in which app in namespace A and app in namespace B may both be using the static asset service to load the same content into different encapsulations in different namespaces.

This approach could mess that up. Thoughts from the other folk?

Comment (bv-jason-dippel) [inline lib/loader/index.js]: How would you feel about defaulting to SHARED_NAMESPACE_NAME, but allow users the option of passing in a specific namespace? Best of both worlds?

Comment (reason-bv) [inline lib/loader/index.js]: Since this is open and not BV-specific, I think I'd rather see it default to not adding to a namespace. E.g. keep a local reference to a hash object in loadedUrls, and only add it to a namespace if provided with one.

File: lib/loader/index.js [hunk 5]
```diff
@@ -14,9 +14,13 @@
...
+var namespace = namespacer.namespace(SHARED_NAMESPACE_NAME);
+namespace.loadedUrls = namespace.loadedUrls || new Array();
```
Comment (reason-bv) [inline lib/loader/index.js]: Why `new Array()` and not `[]`?

File: lib/cssLoadChecker/index.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @fileOverview
+ * Provides functions for checking if style sheet is already loaded. May attempt to load the style
+ * sheet if not present.
```
Comment (reason-bv) [inline lib/cssLoadChecker/index.js]: This is no longer true.

File: lib/loader/index.js [hunk 7]
```diff
@@ -36,6 +38,14 @@ function validateArguments (url, options, callback) {
     throw new Error('`url` must be a string');
   }
 
+  if (options.namespace && typeof options.namespace !== 'string') {
```
Comment (reason-bv) [inline lib/loader/index.js]: To remove confusion, call this namespaceName, perhaps?

File: lib/loader/index.js [hunk 8]
```diff
@@ -77,6 +113,24 @@ module.exports = {
 
     validateArguments(url, options, callback);
 
+    if (options.namespace) {
+      namespace = namespacer.namespace(options.namespace);
```
Comment (reason-bv) [inline lib/loader/index.js]: Better to get a static reference to the namespace early than to run through the function each time. It's a small thing here, since this won't be called frequently. Still.

File: lib/loader/index.js [hunk 9]
```diff
@@ -155,6 +217,9 @@ module.exports = {
    *  to be used as attributes on the created `link` element (optional)
    * @param {number} options.timeout - timeout in milliseconds (defaults to
    *  1000ms)
+   * @param {boolean} options.forceLoad - whether to load the style sheet at the
+   *  specified url or not regardless of whether an earlier attempt to do so has
+   *  already been made
```
Comment (reason-bv) [inline lib/loader/index.js]: Need to add the namespace to the options.

File: lib/loader/index.js [hunk 10]
```diff
@@ -235,6 +318,13 @@ module.exports = {
 
       try {
         injectionNode.appendChild(link);
+
+        if (options.namespace) {
```
Comment (reason-bv) [inline lib/loader/index.js]: You could probably simplify these sorts of if-block by assigning loadedUrls to namespace (and reassign when clearing), so you'd always be able to check the local variable.

**Comment by reason-bv** [comment]
Ok, seems headed in the right direction. Comments made on the latest are all I have, the big picture other than those is ok.

File: lib/loader/index.js (hunk 8 above)
Comment (bv-jason-dippel) [inline lib/loader/index.js]: The reason I put it there as opposed to the top of the function was so that I could include the validation logic within the `validateArguments` function. To me it makes sense to group all that together in the same function, even if it means I get the namespace a little later. If you still disagree and feel like it should be moved to the top of the function, let me know and I'll update it.

File: lib/loader/README.md [hunk 11]
```diff
@@ -41,6 +45,10 @@ attributes on the created `<link>` element.
 placed. The default insertion point is after the first `<script>` element on the page.
 - **timeout**: the time in milliseconds before the loading should be
 considered failed. Defaults to 10000ms.
+- **forceLoad**: a boolean value as to whether the script should
```
Comment (sirugh) [inline lib/loader/README.md]: s/script/stylesheet

File: lib/loader/index.js [hunk 12]
```diff
@@ -47,6 +57,27 @@ function validateArguments (url, options, callback) {
...
+   * we don't need unique URLs for every test, but rather this can be used as a
+   * means of cleaning up.
+   *
+   * @param {string} [namespace] - contains the name of the namespace to clear
+   *  the loadedUrls hash in.
+   */
+  clearLoadedUrls: function (namespaceName) {
+    if (namespaceName) {
+      var namespace = namespacer.namespace(namespaceName);
+      namespace.loadedUrls = [];
```
Comment (sirugh) [inline lib/loader/index.js]: Using namespacer, the proper syntax would be something like

```
try {
 namespace.registerProperty('loadedUrls', []);
} catch (e) {
 // already registered, probably log and do nothing.
}
```

File: lib/loader/index.js [hunk 13]
```diff
@@ -135,6 +186,16 @@ module.exports = {
       var firstScript = getFirstScript();
 
       firstScript.parentNode.insertBefore(script, firstScript);
+
+      _loadedUrls[url] = true;
+
+      if (options.namespaceName) {
+        namespace.loadedUrls = _loadedUrls;
```
Comment (reason-bv) [inline lib/loader/index.js]: Though @sirugh has a point, you should use namespace.registerProperty when first setting a property on the namespace, as the fact that this throws to prevent overwriting is important.

File: test/unit/loader/index.spec.js [hunk 14]
```diff
@@ -197,6 +218,79 @@ describe('lib/loader', function () {
...
+        scriptExecuted = 1;
+      };
+
+      loader.loadScript(url, {
+        namespaceName: TESTING_NAMESPACE_NAME
+      }, function () {
+        expect(scriptExecuted).to.equal(1);
+        expect(namespace.loadedUrls[url]).to.equal(true);
+        done();
+      });
```
Comment (sirugh) [inline test/unit/loader/index.spec.js]: This function should attempt to load the same url a second time and assert that it is the same as the loaded script from the first execution of `loadScript`.

Comment (sirugh) [inline test/unit/loader/index.spec.js]: The same goes for the stylesheet check in the same vein.

File: lib/loader/index.js [hunk 15]
```diff
@@ -47,6 +57,27 @@ function validateArguments (url, options, callback) {
 
 module.exports = {
   /**
+   * Clears the loaded URLs
+   *
+   * Clears any URLs that might have been stored in the loadedUrls hash, either
+   * locally or in the given namespace(optional). This is needed for testing so
```
Comment (sirugh) [inline lib/loader/index.js]: If the only reason this exists is for testing, it probably doesn't need to exist, at least not in the module. Is there another way to get a handle to an internal data structure and clear it via tests? You can get a handle on namespace.loadedUrls but not the internal loadedUrls right?

File: lib/loader/index.js (hunk 13 above)
Comment (sirugh) [inline lib/loader/index.js]: I deleted that comment because I realized it was not relevant on the given function, but yes.

File: lib/loader/index.js (hunk 15 above)
Comment (bv-jason-dippel) [inline lib/loader/index.js]: That is correct. During testing, I'm able to access namespace.loadedUrls but not the internal loadedUrls because I'm not exporting it. The best way I've thought of doing it is with this function, but as you said if it's only purpose is for testing, it would be nice to remove it.

I don't think there's a need for it in actual execution, but I've not had any luck removing it and still being able to test things nicely. I'm open to suggestions.

Comment (reason-bv) [inline lib/loader/index.js]: I'm not adverse to exposing things for test purposes.

Comment (sirugh) [inline lib/loader/index.js]: If you absolutely must, then I suggest marking it as private `_clearLoadedUrls` or `_cleanup` or something that indicates it's not part of the public API of this module.

File: lib/loader/index.js (hunk 13 above)
Comment (bv-jason-dippel) [inline lib/loader/index.js]: @sirugh, @reason-bv Just to be clear, you guys are talking about something like...
```js
_loadedUrls[url] = true;

if (options.namespaceName) {
  if (namespace.loadedUrls) {
    namespace.loadedUrls = _loadedUrls;
  }
  else {
    namespace.registerProperty('loadedUrls', _loadedUrls);
  }
}
else {
  loadedUrls = _loadedUrls;
}
```

Comment (reason-bv) [inline lib/loader/index.js]: Not quite. The first time you set that property on the namespace, you have to register it. Not afterwards.

So it might be better for you to, at the outset and once, register an object on the namespace, then add loadedUrls to that object as you need to elsewhere in the code.

Comment (tnunamak) [inline lib/loader/index.js]: As I understand it, by calling `registerProperty`, you are establishing ownership over that property. Multiple apps can read it, but only one should call `registerProperty` and that is the only one that should be allowed to modify it. Correct?

Comment (sirugh) [inline lib/loader/index.js]: They can de-reference it or assign properties of the property. `registerProperty` is just a helper that throws an error if it happens to exist already. So if you are performing possible destructive ops you can use it to be safe.

Comment (tnunamak) [inline lib/loader/index.js]: I've never used [`set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set) before, could we use it to enforce this? Could you determine if the property owner is the one making a modification by inspecting the context?

Comment (tnunamak) [inline lib/loader/index.js]: > They can de-reference it or assign properties of the property.

Assign properties as in modify its value? This seems like a risky thing to allow.

Comment (reason-bv) [inline lib/loader/index.js]: What @sirugh said - use of registerProperty is just a way to ensure that when there are conflicts, there will be a clear exception rather than confusion as two apps edit each other's data haphazardly.

I don't think there is any intent of having app2 read the data used by app1.

Comment (tnunamak) [inline lib/loader/index.js]: Okay. As long as we're not advocating for namespace properties to be shared, this seems fine. If/when we do want apps to share state, we will want to use namespace somehow, won't we?

Comment (reason-bv) [inline lib/loader/index.js]: Yes. 

In fact what is going on here is sharing of properties between the loaders of different apps using the same namespace.

Which means I'm being dumb about what I'm suggesting with registerProperty. The same code running against the same namespace in different apps was one of the original goals here. So hmm. Don't use register property. Do write some inline documentation to explain why you are not using it.

File: lib/loader/index.js (hunk 15 above)
Comment (taylormck) [inline lib/loader/index.js]: It's possible to write tests that would require this, but as it's just a "clean" method, it's fine by me to have around.

Comment (sirugh) [inline lib/loader/index.js]: Then I'm fine having it `_clearLoadedUrls`.

File: lib/cssLoadChecker/README.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
+# CSS Load Checker
+
+This module provides a means of checking if a CSS file has been loaded already or not. It does this by adding a test div to the page and checking the computed styles of it for the `display: none` style.
+
+```js
+var cssLoadChecker = require('bv-ui-core/lib/cssLoadChecker');
+
+if (cssLoadChecker.checkForCss('bv-verify-css-loaded')) {
```
Comment (sirugh) [inline lib/cssLoadChecker/README.md]: 1. Should this be a promise? Technically the promise only needs to be resolved once as once css loads, it's "loaded" and can't be unloaded. Am I just overengineering? @savetheclocktower? @tnunamak?

```javascript
cssLoadChecker.cssIsLoaded('bv-verify-css-loaded')
  .then(() => { 
    // Continue with your life.
  }
  .fail(() => {
     loader.loadStyleSheet(namespace.cssUrl, callback)
  })
```

2. This is personal preference: `checkForCss` -> `isCssLoaded` since it returns a boolean.

Comment (tnunamak) [inline lib/cssLoadChecker/README.md]: If it's not an async check, using a promise does feel like overengineering to me.

Comment (reason-bv) [inline lib/cssLoadChecker/README.md]: Promises would require a library, and we're trying to keep to zero dependencies here.

Comment (sirugh) [inline lib/cssLoadChecker/README.md]: Oh. Right. Duh. Thanks for the input guys. Carry on.

File: lib/loader/index.js [hunk 17]
```diff
@@ -77,6 +116,26 @@ module.exports = {
...
+     * The goal of using the namespacer is to allow multiple apps on the same
+     * page to share the properties within the namespace, namely loadedUrls. For
+     * this reason, we are forgoing the use of registerProperty to instantiate
+     * the list of loaded URLs. This will allow the loader module to be run in
+     * different apps but access the same namespace property.
+     */
+
+    if (options.namespaceName) {
+      namespace = namespacer.namespace(options.namespaceName);
+      _loadedUrls = namespace.loadedUrls || [];
```
Comment (sirugh) [inline lib/loader/index.js]: Default to object. Cannot reference array index by string.

File: lib/loader/index.js [hunk 18]
```diff
@@ -235,6 +332,15 @@ module.exports = {
 
       try {
         injectionNode.appendChild(link);
+
```
Comment (sirugh) [inline lib/loader/index.js]: If `appendChild` and the resulting script inject is synchronous, this is fine, but if it isn't you may consider putting 336-343 in the `onload`.

**Comment by sirugh** [comment]
LGTM.
